lib/state.js

```diff
@@ -0,0 +1,66 @@
+const debug = require('debug')('snyk:nodejs:state');
+
+const state = {
+  events: [],
+  filters: {},
+  packages: {},
+};
+
+module.exports = {
+  get,
+  clean,
+  addEvent,
+  addFilter,
+  addPackage,
+  removeFilter,
+};
+
+function get() {
+  return state;
+}
+
+function clean() {
+  state.events.length = 0;
+}
+
+function addEvent(event) {
+  event.timestamp = (new Date()).toISOString();
+  state.events.push(event);
+  debug(`Event added to transmission queue: ${JSON.stringify(event)}`);
+}
+
+function addFilter(packageName, fileRelativePath, functionName) {
+  const filters = state.filters;
+
+  if (!(packageName in filters)) {
+    filters[packageName] = {};
+  }
+  if (!(fileRelativePath in filters[packageName])) {
+    filters[packageName][fileRelativePath] = {};
+  }
+  filters[packageName][fileRelativePath][functionName] = null;
+}
+
+function removeFilter(packageName, fileRelativePath, functionName) {
+  const filters = state.filters;
+
+  delete filters[packageName][fileRelativePath][functionName];
+  if (Object.keys(filters[packageName][fileRelativePath]).length === 0) {
+    delete filters[packageName][fileRelativePath];
+  }
+  if (Object.keys(filters[packageName]).length === 0) {
+    delete filters[packageName];
+  }
+}
+
+function addPackage(name, version) {
+  if (!(name in state.packages)) {
+    state.packages[name] = {
+      versions: [],
```

Comment:
Do you think it should also be a map for future support? (like the "introduced through")

Reply:
not sure, I hadn't given it as much thought as you had.
what do you have in mind? how would this object look like, (if and) when we're gathering the "introduced through" data?